schema/src/lib.rs

```diff
@@ -1,52 +1,105 @@
 //! Module for schematizing rust types in other languages for translation.
 
 #![no_std]
+#![allow(unsafe_code)]
 
 extern crate alloc;
 
+mod serialize;
+
 use alloc::{
+    borrow::ToOwned as _,
     boxed::Box,
-    collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+    collections::{btree_map, btree_set},
     format,
     string::String,
     vec,
     vec::Vec,
 };
-use core::num::NonZeroU8;
 
 /// Derive schema. It will make your structure schemaable
-pub use iroha_schema_derive::IntoSchema;
+pub use iroha_schema_derive::*;
 use serde::Serialize;
 
-/// Metadata map
-pub type MetaMap = BTreeMap<String, Metadata>;
+/// Helper struct for building a full schema
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MetaMap(pub(crate) btree_map::BTreeMap<core::any::TypeId, (String, Metadata)>);
+
+impl PartialEq<btree_map::BTreeMap<core::any::TypeId, (String, Metadata)>> for MetaMap {
+    fn eq(&self, other: &btree_map::BTreeMap<core::any::TypeId, (String, Metadata)>) -> bool {
+        self.0.eq(other)
+    }
+}
+
+impl MetaMap {
+    fn key<K: 'static>() -> core::any::TypeId {
+        core::any::TypeId::of::<K>()
+    }
+
+    /// Create new [`Self`]
+    pub const fn new() -> MetaMap {
+        Self(btree_map::BTreeMap::new())
+    }
+    /// Return `true` if the map contains a metadata for the specified [`core::any::TypeId`]
+    pub fn contains_key<K: 'static>(&self) -> bool {
+        self.0.contains_key(&Self::key::<K>())
+    }
+    /// Insert a key-value pair into the map.
+    pub fn insert<K: IntoSchema>(&mut self, v: Metadata) -> bool {
+        self.0
+            .insert(Self::key::<K>(), (K::type_name(), v))
+            .is_none()
+    }
+    /// Return a reference to the value corresponding to the [`core::any::TypeId`] of the schema type
+    pub fn get<K: 'static>(&self) -> Option<&Metadata> {
+        self.0.get(&Self::key::<K>()).map(|(_, schema)| schema)
+    }
+}
+
+impl IntoIterator for MetaMap {
+    type Item = (core::any::TypeId, (String, Metadata));
+    type IntoIter = btree_map::IntoIter<core::any::TypeId, (String, Metadata)>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+/// Globally unique type identifier
+///
+/// No critical code should rely on this trait unless a test
+/// is devised that can prove that all impls are unique
+pub trait TypeId: 'static {
+    /// Return unique type id
+    // TODO: Should return &str or ConstString.
+    fn id() -> String;
+}
 
 /// `IntoSchema` trait
-pub trait IntoSchema {
-    /// Returns unique type name.
-    // TODO: Should return &str if possible or be immutable string
+pub trait IntoSchema: TypeId {
+    /// Name under which a type is represented in the schema
+    // TODO: Should return &str or ConstString.
     fn type_name() -> String;
```

Comment:
hmm, ok, shouldn't than `IntoSchema::type_name` had default implementation which uses `TypeId::id` to avoid duplication in few places?

Reply:
I thought so as well, but no because of generic parameters. If you look at `Box<SignatureWrapperOf<T>>`, by your suggestion `IntoSchema::type_name` would end up being `Box<SignatureWrapperOf<T>` which is incorrect because it should to be `Box<SignatureOf<T>>`

To illustrate (if I used `TypeId::id` as the default):
```rs
trait TypeId {
    fn type_name() -> String {
        // NOTE: this would return `format!("Box<{}>", T::id())` 
        // but it should return `format!("Box<{}>", T::type_name())
        T::id()
    }
}
impl<T: TypeId> TypeId for Box<T> {
    fn id() -> String {
        format!("Box<{}>", T::id())
    }
}
impl<T: IntoSchema> IntoSchema for Box<T> {
    // ...
}
```